File: problem_047.c
// -*- compile-command: "gcc -o problem_047 problem_047.c ceuler.c -Wall -lm" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include <stdlib.h>
#include "ceuler.h"

int primeFactors(int number, const int * primes, const int primeIndex)
{
    int count = 0;
    int factors = 1;
    int pIndex = 0;

    while (number > 1)
    {
        if (count == 1) factors++;
        if (number % primes[pIndex] == 0)
        {
            number /= primes[pIndex];
            count++;
        }
        else
        {
            pIndex++;
            count = 0;
        }
    }
    return factors;
}

int main()
{
    printf("Project Euler - Problem 47:\n"
           "What is the of four consecutive integers to have four distinct prime factors?\n\n");

    int primes[50000] = {};
    int primeIndex = 0;
    int composite = 1;
    int count = 0, candidate = 0;

    while ( composite++ && count != 4 )
    {
        if (isPrimeOpt(composite, primes) == true)
        {
            primes[primeIndex++] = composite;
            count = 0;
        }
        else // is composite number
        {
            if (primeFactors(composite, primes, primeIndex) == 4)
            {
                if (count == 0)
                    candidate = composite;
                count++;
            }
            else
            {
                count = 0;
            }
        }
    }

    printf("First digit: %d\n", candidate);

    return 0;
}

File: problem_047.rs
// -*- compile-command: "rustc -o problem_047_rs problem_047.rs" -*-
// Copyright (c) 2016 Michael Caldwell

fn main() {
    println!("Project Euler - Problem 47:");
    println!("What is the first of four consecutive integers to have four distinct prime factors?\n");

    let mut primes: [usize; 1000000] = [0; 1000000];
    let mut count = 0;

    for n in 2..primes.len() {
        // Prime seive
        if primes[n] == 0 {
            let mut m = 2;
            let mut i = m * n;
            while i < primes.len() {
                primes[i] += 1;
                m += 1;
                i = m * n;
            }
        }

        // 4 prime factors
        if primes[n] == 4 {
            count += 1;
            if count == 4 {
                println!("{}", n-3);
                break;
            }
        } else {
            count = 0;
        }
    }
}
